sdk/storage/Azure.Storage.Blobs.Batching/src/BlobBatch.cs

```diff
@@ -0,0 +1,294 @@
+{
+    /// <summary>
+    /// A <see cref="BlobBatch"/> allows you to batch multiple Azure Storage
+    /// operations in a single request via <see cref="BlobBatchClient.SubmitBatch"/>.
+    ///
+    /// For more information, see
+    /// <see href="https://docs.microsoft.com/en-us/rest/api/storageservices/blob-batch" />.
+    /// </summary>
+    public class BlobBatch
+    {
+        /// <summary>
+        /// The <see cref="BlobBatchClient"/> associated with this batch.  It
+        /// provides the Uri, BatchOperationPipeline, etc.
+        /// </summary>
+        private readonly BlobBatchClient _client;
+
+        /// <summary>
+        /// Storage requires each batch request to contain the same type of
+        /// operation.
+        /// </summary>
+        private BlobBatchOperationType? _operationType = null;
+
+        /// <summary>
+        /// The list of messages that will be sent as part of this batch.
+        /// </summary>
+        private readonly IList<HttpPipelineMessage> _messages = new List<HttpPipelineMessage>();
+
+        /// <summary>
+        /// A value indicating whether the batch has already been submitted.
+        /// </summary>
+        internal bool Submitted { get; private set; } = false;
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="BlobBatch"/> for mocking.
+        /// </summary>
+        protected BlobBatch()
+        {
+        }
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="BlobBatch"/> class.
+        /// </summary>
+        /// <param name="client">
+        /// The <see cref="BlobBatchClient"/> associated with this batch.
+        /// </param>
+        public BlobBatch(BlobBatchClient client) =>
+            _client = client ?? throw new ArgumentNullException(nameof(client));
+
+        /// <summary>
+        /// Gets the list of messages to submit as part of this batch.
+        ///
+        /// Note that calling this method marks the batch as submitted and no
+        /// further operations can be added.
+        /// </summary>
+        /// <returns>
+        /// The list of messages to submit as part of this batch.
+        /// </returns>
+        internal IList<HttpPipelineMessage> GetMessagesToSubmit()
+        {
+            Submitted = true;
+            return _messages;
+        }
+
+        /// <summary>
+        /// Verify whether the <paramref name="client"/> is the batch client
+        /// associated with this batch.
+        /// </summary>
+        /// <param name="client">The BlobBatchClient to check.</param>
+        internal bool IsAssociatedClient(BlobBatchClient client) =>
+            _client == client;
+
+        /// <summary>
+        /// Verify that the desired operation is allowed.
+        /// </summary>
+        /// <param name="operationType">
+        /// The type of operation to perform.
+        /// </param>
+        private void EnsureAllowed(BlobBatchOperationType operationType)
```

Comment:
Consider renaming to ThrowIfNotAllowed.

Reply:
It does more than validation.  Changed name to `SetBatchOperationType`.